remote_attestation/rust/src/attestation/unattested_session.rs

```diff
@@ -0,0 +1,146 @@
+        Self {
+            tee_certificate: tee_certificate.to_vec(),
+            expected_tee_measurement: expected_tee_measurement.to_vec(),
+        }
+    }
+}
+
+pub trait UnattestedSession {
+    fn get_attestation_info(&self, public_key_hash: &[u8]) -> Option<AttestationInfo>;
+    fn verify_attestation(self, peer_identity: &AttestationIdentity) -> anyhow::Result<()>;
+}
+
+impl UnattestedSession for UnattestedSelf {
+    fn get_attestation_info(&self, public_key_hash: &[u8]) -> Option<AttestationInfo> {
+        // Generate attestation info with a TEE report.
+        // TEE report contains a hash of the public key.
+        Some(AttestationInfo {
+            report: Some(AttestationReport::new(&public_key_hash)),
+            certificate: self.tee_certificate.clone(),
+        })
+    }
+
+    fn verify_attestation(self, _peer_identity: &AttestationIdentity) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+impl UnattestedSession for UnattestedPeer {
+    fn get_attestation_info(&self, _public_key_hash: &[u8]) -> Option<AttestationInfo> {
+        None
+    }
+
+    fn verify_attestation(self, peer_identity: &AttestationIdentity) -> anyhow::Result<()> {
+        verify_attestation(&peer_identity, &self.expected_tee_measurement)
+            .context("Couldn't verify peer attestation info")
+    }
+}
+
+impl UnattestedSession for MutuallyUnattested {
+    fn get_attestation_info(&self, public_key_hash: &[u8]) -> Option<AttestationInfo> {
+        // Generate attestation info with a TEE report.
+        // TEE report contains a hash of the public key.
+        Some(AttestationInfo {
+            report: Some(AttestationReport::new(&public_key_hash)),
+            certificate: self.tee_certificate.clone(),
+        })
+    }
+
+    fn verify_attestation(self, peer_identity: &AttestationIdentity) -> anyhow::Result<()> {
+        verify_attestation(&peer_identity, &self.expected_tee_measurement)
+            .context("Couldn't verify peer attestation info")
+    }
+}
+
+/// Verifies the validity of the attestation info:
+/// - Checks that the TEE report is signed by TEE Provider’s root key.
+/// - Checks that the public key hash from the TEE report is equal to the hash of the public key
+///   presented in the server response.
+/// - Extracts the TEE measurement from the TEE report and compares it to the
+///   `expected_tee_measurement`.
+fn verify_attestation(
+    peer_identity: &AttestationIdentity,
+    expected_tee_measurement: &[u8],
+) -> anyhow::Result<()> {
+    // Verify peer attestation info.
+    // TODO(#1867): Add remote attestation support, use real TEE reports and check that
+    // `AttestationInfo::certificate` is signed by one of the root certificates.
+    let peer_attestation_info = peer_identity
+        .attestation_info
+        .as_ref()
+        .context("Peer identity doesn't contain attestation info")?;
+    peer_attestation_info
+        .verify()
```

Comment:
I'm not sure an attestation info (whatever it ends up being) can "verify itself". Should it not be verified by the caller by checking its signature? I know there is a TODO, but this verify method here seems confusing to me, or maybe I'm missing something.

Reply:
Attestation info has both report and corresponding provider's certificate in it.
So it can check the validity of the report without additional arguments.

Reply:
What's the point of that? The peer can just make up a key and use that to sign whatever it wants, but that's meaningless. We need to check that it was ultimately signed by a cert / key that the client trusts, not just whatever the peer sends over, right?